packages/extension/src/background/handlers/State.ts

```diff
@@ -213,6 +222,40 @@ export default class State {
     return this.#signRequests[id];
   }
 
+  public rpcSend (request: RequestRpcSend): Promise<JsonRpcResponse> {
+    assert(this.#provider, 'Cannot call pub(rpc.subscribe) before provider has been set');
+
+    return this.#provider.send(request.method, request.params);
+  }
+
+  // Change the provider the extension is exposing
+  public rpcSetProvider (provider: RequestRpcProvider): Promise<null> {
+    switch (provider.type) {
+      case 'WsProvider': {
+        this.#provider = new WsProvider(provider.payload);
+        break;
+      }
+      default: {
+        return Promise.reject(new Error(`Unable to instantiate provider of type ${provider.type}`));
+      }
+    }
+
+    return Promise.resolve(null);
+  }
+
+  public rpcSubscribe (request: RequestRpcSubscribe, cb: ProviderInterfaceCallback): Promise<number> {
+    assert(this.#provider, 'Cannot call pub(rpc.subscribe) before provider has been set');
+
+    return this.#provider.subscribe(
+      request.type,
+      request.method,
+      request.params,
+      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+      // @ts-ignore WsProvider gives me (error, result)=>void, whereas (result)=>void is expected
+      (_error, res) => { cb(res); }
```

Comment:
this is weird... ProviderInterface's `subscribe` method has signature `(result)=>void`, but somehow with WsProvider I get `error` (null most of the time) as 1st arg

Reply:
ah, yes, I remember bumping into this issue. the signature in WsProvider is wrong. you're right, [`subscribe`](https://github.com/polkadot-js/api/blob/333ff8528b790eb790e5d31f951b01ed35c3c746/packages/rpc-provider/src/ws/Provider.ts#L236) takes a [`ProviderInterfaceCallback`](https://github.com/polkadot-js/api/blob/333ff8528b790eb790e5d31f951b01ed35c3c746/packages/rpc-provider/src/types.ts#L39) of signature `(result)=>void` but the callback is actually called [this way](https://github.com/polkadot-js/api/blob/333ff8528b790eb790e5d31f951b01ed35c3c746/packages/rpc-provider/src/ws/Provider.ts#L352-L357), i.e. with `(_error, res)`. I guess the ProviderInterfaceCallback signature should be changed